### ISSUE TITLE
Query understanding SLM: replace traditional NLP with a single model

### ISSUE THREAD

**Issue by anibjoshi** [stratalab/strata-core#2261]
## Context

Parent: #1584 (BM25 gaps)
Replaces separate NLP features: #2249 (fuzzy), #2251 (multi-lang stemming), #2253 (synonyms), #2257 (phonetic), #2259 (spell check)

## Problem

Traditional search engines (Elasticsearch, Redis Search) ship dozens of separate NLP modules: stemmers per language, synonym dictionaries, phonetic encoders, fuzzy matchers, spell checkers, stop word lists. Each is a narrow heuristic that handles one failure mode. Strata has a built-in inference engine — we should use it.

## Insight

One small language model in the query path can replace all of these:
- **Spelling correction**: "macine lerning" → "machine learning"
- **Synonym expansion**: "car" → "car automobile vehicle"
- **Language normalization**: "laufend" (German) → "running"
- **Phonetic resolution**: "Steven" → "Steven Stephen"
- **Intent detection**: "how to fix error 404" → extract key terms + classify intent
- **Query reformulation**: "papers about ML published last year" → structured query

All in a single forward pass. One model replaces 100 NLP hacks.

### 1. Query understanding pipeline

Insert an SLM between user query and BM25/vector retrieval:
```
user query → [Query Understanding SLM] → rewritten query → BM25 + vector
```

### 2. Latency budget

The SLM must complete in <50ms to stay within the search budget (5s total, but query rewriting is on the critical path before retrieval). Options:
- **Quantized SLM** (Q4/Q8) running on Strata's built-in inference engine
- **Distilled model** fine-tuned specifically for query understanding
- **Cached rewrites** for common query patterns

### 3. Training data

Fine-tune on:
- Misspelling → correction pairs (from search logs)
- Query → expanded query pairs (from click-through data)
- Query → intent classification
- Multi-language query → English normalized form

### 4. Model candidates

- Fine-tuned Phi-3.5-mini or TinyLlama for query rewriting
- Smaller: fine-tuned BERT/DistilBERT for classification + token-level correction
- Smallest: character-level CNN for spelling correction only (sub-1ms)

### 5. Fallback

When the SLM is unavailable or too slow, fall back to raw query (current behavior). The engine must work without the model — the SLM is a quality boost, not a hard dependency.

## Architecture Decision

This is the Strata way: instead of shipping 15 separate NLP modules with configuration knobs, ship one model that learns the mapping. The inference engine is already built (#strata-inference). The recipe system already supports expansion. The missing piece is a fast, specialized SLM.

## Relationship to existing issues

- #2249 (fuzzy matching): SLM handles this via spelling correction
- #2251 (multi-lang stemming): SLM handles this via language normalization  
- #2253 (synonyms): SLM handles this via query expansion
- #2257 (phonetic): SLM handles this via entity normalization
- #2259 (spell check): SLM handles this directly

These issues remain open as fallback implementations if the SLM approach proves too slow for certain use cases (e.g., <1ms latency requirement). But the primary path should be the SLM.

## Acceptance Criteria

- [ ] Query understanding SLM integrated into search pipeline
- [ ] Latency < 50ms for query rewriting (p99)
- [ ] Handles: spelling, synonyms, basic normalization
- [ ] Fallback to raw query when model unavailable
- [ ] Measurable nDCG improvement on BEIR with misspelled/variant queries
- [ ] Recipe knob to enable/disable
